src/tools/softhsm/migrate.rs

```diff
@@ -0,0 +1,1068 @@
+        Ok(_) => (),
+        Err(e) => return Err(format!("Failed to create object: {}", e)),
+    }
+
+    Ok(())
+}
+
+fn import_key(
+    pkcs11: &mut FuncList,
+    class: CK_OBJECT_CLASS,
+    attrs: &mut Vec<Attribute>,
+    ekey: &Vec<u8>,
+) -> Result<(), String> {
+    let key_type = find_ulong(attrs, CKA_KEY_TYPE);
+    if key_type == CK_UNAVAILABLE_INFORMATION {
+        return Err("Key type not found".to_string());
+    }
+    match key_type {
+        CKK_RSA | CKK_EC | CKK_GENERIC_SECRET | CKK_AES | CKK_SHA_1_HMAC
+        | CKK_SHA256_HMAC | CKK_SHA384_HMAC | CKK_SHA512_HMAC
+        | CKK_SHA224_HMAC | CKK_SHA3_224_HMAC | CKK_SHA3_256_HMAC
+        | CKK_SHA3_384_HMAC | CKK_SHA3_512_HMAC | CKK_EC_EDWARDS
+        | CKK_EC_MONTGOMERY | CKK_HKDF => (),
+        _ => return Err(format!("Unsupported key type {:x}", key_type)),
+    }
+
+    for a in &mut *attrs {
+        /* Skip values that are not byte buffers but can be larger than 16 bytes */
+        match a.attr.type_ {
+            CKA_ALLOWED_MECHANISMS => continue,
+            _ => (),
+        }
+        /* Assume all other values bigger than 16 bytes (IV size) are encrypted */
+        if a.attr.ulValueLen > 16 {
+            match decrypt_cbc_buffer(pkcs11, a.value.as_slice(), ekey) {
+                Ok(v) => {
+                    /* replace encrypted with plain text */
+                    a.value = v;
+                    a.attr.pValue = a.value.as_ptr() as *const _ as CK_VOID_PTR;
+                    a.attr.ulValueLen = a.value.len() as CK_ULONG;
+                }
+                Err(e) => {
+                    return Err(format!("Failed to decrypt key values: {}", e));
+                }
+            }
+        }
+    }
+    import_obj(pkcs11, class, attrs)
+}
+
+const CKA_VENDOR_SOFTHSM: CK_ULONG = CKA_VENDOR_DEFINED + 0x5348;
+/*
+const CKA_OS_TOKENLABEL: CK_ULONG = CKA_VENDOR_SOFTHSM + 1;
+const CKA_OS_TOKENSERIAL: CK_ULONG = CKA_VENDOR_SOFTHSM + 2;
+const CKA_OS_TOKENFLAGS: CK_ULONG = CKA_VENDOR_SOFTHSM + 3;
+const CKA_OS_SOPIN: CK_ULONG = CKA_VENDOR_SOFTHSM + 4;
+*/
+const CKA_OS_USERPIN: CK_ULONG = CKA_VENDOR_SOFTHSM + 5;
+
+fn migrate(pkcs11: &mut FuncList, token: &PathBuf, pin: &Vec<u8>) -> u8 {
+    /* check this is a token dir */
+    let mut tokfile = token.clone();
+    tokfile.push("token.object");
+
+    let attrs = match read_object(&tokfile) {
+        Ok(a) => a,
+        Err(e) => {
+            eprintln!("{}", e);
+            return 0xFE;
+        }
+    };
+
+    let pin_attr = match attrs.iter().find(|a| a.attr.type_ == CKA_OS_USERPIN) {
+        Some(a) => a,
+        None => {
+            eprintln!("Encryption key not found!");
+            return 0xFE;
+        }
+    };
+
+    let encryptionkey = match encryption_key(pkcs11, &pin_attr.value, pin) {
+        Ok(k) => k,
+        Err(e) => {
+            eprintln!("Failed to get key: {}", e);
+            return 0xFE;
+        }
+    };
+
+    let mut error = 0u8;
```

Comment:
we can have only 256 errors? Is it enough?

Reply:
plenty for now